opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java

```diff
@@ -84,43 +82,70 @@ public static Tracer get() {
+     * @param provider Provider for the tracer to use as global tracer.
+     * @return {@code true} if the provided tracer was registered as a result of this call,
+     * {@code false} otherwise.
      */
-    public static synchronized void register(final Tracer tracer) {
-        if (tracer == null) {
-            throw new NullPointerException("Cannot register GlobalTracer <null>.");
-        }
-        if (tracer instanceof GlobalTracer) {
-            LOGGER.log(Level.FINE, "Attempted to register the GlobalTracer as delegate of itself.");
-            return; // no-op
-        }
-        if (isRegistered() && !GlobalTracer.tracer.equals(tracer)) {
-            throw new IllegalStateException("There is already a current global Tracer registered.");
+    public static synchronized boolean registerIfAbsent(final Callable<Tracer> provider) {
+        requireNonNull(provider, "Cannot register GlobalTracer from provider <null>.");
+        if (!isRegistered()) try {
+            final Tracer suppliedTracer = requireNonNull(provider.call(), "Cannot register GlobalTracer <null>.");
+            if (!(suppliedTracer instanceof GlobalTracer)) {
+                GlobalTracer.tracer = suppliedTracer;
+                return true;
+            }
+        } catch (RuntimeException rte) {
+            throw rte; // Re-throw as-is
+        } catch (Exception ex) {
+            throw new IllegalStateException("Exception obtaining tracer from provider: " + ex.getMessage(), ex);
         }
-        GlobalTracer.tracer = tracer;
+        return false;
     }
 
     /**
-     * Identify whether a {@link Tracer} has previously been registered.
+     * Register a {@link Tracer} to back the behaviour of the {@link #get() global tracer}.
      * <p>
-     * This check is useful in scenarios where more than one component may be responsible
-     * for registering a tracer. For example, when using a Java Agent, it will need to determine
-     * if the application has already registered a tracer, and if not attempt to resolve and
-     * register one itself.
+     * Registration is a one-time operation, attempting to call it more often will result in a runtime exception.
+     * <p>
+     * Every application intending to use the global tracer is responsible for registering it once
+     * during its initialization.
      *
-     * @return Whether a tracer has been registered
+     * @param tracer Tracer to use as global tracer.
+     * @throws RuntimeException if there is already a current tracer registered
+     * @see #registerIfAbsent(Callable)
+     * @deprecated Please use 'registerIfAbsent' instead which does not attempt a double registration.
      */
-    public static synchronized boolean isRegistered() {
-        return !(GlobalTracer.tracer instanceof NoopTracer);
+    @Deprecated
+    public static void register(final Tracer tracer) {
+        if (!registerIfAbsent(provide(tracer))
+                && !tracer.equals(GlobalTracer.tracer)
```

Comment:
I am not sure if these checks are mandatory here. `registerIfAbsent(provide(tracer)` is called before and should contain the same checks and throw the same exception.

Reply:
@pavolloffay These checks are necessary to keep the current behaviour backward-compatible.

This method is deprecated and must be removed in a future version. Until then, the error-behaviour must not change.